execution_engine2.spec

```diff
@@ -190,43 +216,120 @@ module execution_engine2 {
+            error - job is no longer running, but failed with an error
+            terminated - job is no longer running, terminated either due to user cancellation,
+                         admin cancellation, or some automated task
+        error_code - int - internal reason why the job is an error. one of the following:
+            0 - unknown
+            1 - job crashed
+            2 - job terminated by automation
+            3 - job ran over time limit
+            4 - job was missing its automated output document
+            5 - job authentication token expired
+        errormsg - string - message (e.g. stacktrace) accompanying an errored job
+
+        terminated_code - int - internal reason why a job was terminated, one of:
+            0 - user cancellation
+            1 - admin cancellation
+            2 - terminated by some automatic process
+
+        @optional error_code
+        @optional errormsg
+        @optional terminated_code
+        @optional estimating
+        @optional running
+        @optional finished
+    */
+
+
+    typedef structure {
+        job_id _id;
+        string user;
+        string authstrat;
+        int wsid;
+        string status;
+        RunJobParams job_input;
+        timestamp created;
+        timestamp updated;
+
+        timestamp estimating
+        timestamp running;
+        timestamp finished;
+
+        int error_code;
+        string errormsg;
+
+        int terminated_code;
+    } JobState;
+
     /*
         get current status of a job
     */
-    funcdef check_job(CheckJobParams params) returns (UnspecifiedObject job_state) authentication required;
+    funcdef check_job(CheckJobParams params) returns (JobState job_state) authentication required;
 
     /*
         job_states - states of jobs
     */
     typedef structure {
-        mapping<job_id, UnspecifiedObject> job_states;
+        mapping<job_id, JobState> job_states;
     } CheckJobsResults;
 
+    /*
+        As in check_job, projection strings can be used to return only useful fields.
+        see CheckJobParams for allowed strings.
+    */
     typedef structure {
         list<job_id> job_ids;
         list<string> projection;
     } CheckJobsParams;
 
     funcdef check_jobs(CheckJobsParams params) returns (CheckJobsResults) authentication required;
+
     /*
-      Check job for all jobs in a given workspace
+        Check status of all jobs in a given workspace. Only checks jobs that have been associated
+        with a workspace at their creation.
     */
     typedef structure {
         string workspace_id;
         list<string> projection;
     } CheckWorkspaceJobsParams;
     funcdef check_workspace_jobs(CheckWorkspaceJobsParams params) returns (CheckJobsResults) authentication required;
+
     typedef structure {
         job_id job_id;
     } CancelJobParams;
+
+    /*
+        Cancels a job. This results in the status becoming "terminated" with termination_code 0.
```

Comment:
This is true right now. However, we should add a termination_code parameter so we can terminate it with

```
    terminated_by_user = 0
    terminated_by_admin = 1
    terminated_by_automation = 2
```

Should we update the documentation to say this now, or update it once that functionality is actually put in?

Reply:
Let's try to keep the doc up to date with the code.